# Cleaning `laptops.csv' dataset from DQ





## Import libraries and data

In [0]:
# imports
import pandas as pd


# read csv file to namespace as pd.dataframe object
laptops = pd.read_csv('/content/drive/My Drive/Datasets/laptops.csv', encoding='Latin-1', header=0)


## Explore dataset

In [2]:
# explore dataframe using df.head() and df.info() methods and printing first row of dataframe
print(laptops.head(10))
print()
laptops.info() #df.info() does not return a value 
print()
print(laptops.iloc[0])


  Manufacturer       Model Name  ...  Weight Price (Euros)
0        Apple      MacBook Pro  ...  1.37kg       1339,69
1        Apple      Macbook Air  ...  1.34kg        898,94
2           HP           250 G6  ...  1.86kg        575,00
3        Apple      MacBook Pro  ...  1.83kg       2537,45
4        Apple      MacBook Pro  ...  1.37kg       1803,60
5         Acer         Aspire 3  ...   2.1kg        400,00
6        Apple      MacBook Pro  ...  2.04kg       2139,97
7        Apple      Macbook Air  ...  1.34kg       1158,70
8         Asus  ZenBook UX430UN  ...   1.3kg       1495,00
9         Acer          Swift 3  ...   1.6kg        770,00

[10 rows x 13 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Manufacturer              1303 non-null   object
 1   Model Name                1303 non-null   object
 2 

Observations:

* There are 1303 rows in the dataframe. We observe `Operating System Version` column contains 170 null entries (`NaN`).
* All columns have `object` dtype where clearly some could more usefully take numeric dtypes
* Column names have mixed casing (title case or upper) and contain whitespace between words rather than underscores
* Some column names contain unneccesary whitespace we should strip
* Some column names contain parentheses 
* Some column names are too verbose e.g `Operating System` and can be abbreviated for efficiency

First we clean the column names:


In [3]:
# define a function clean_col_name to be applied to each column name string to remove observed inconsistencies and inefficiencies
def clean_col_name(col_name):
  cleaned_col_name = (col_name.
              lower().
              strip().
              replace('(','').
              replace(')','').
              replace(' ','_').
              replace('operating_system', 'os')
  )
  return cleaned_col_name

# use iterable df.columns attribute to apply clean_col_name function to each column name in laptops df and assign back to laptops.columns
laptops.columns = [clean_col_name(c) for c in laptops.columns]
print(laptops.columns)



Index(['manufacturer', 'model_name', 'category', 'screen_size', 'screen',
       'cpu', 'ram', 'storage', 'gpu', 'os', 'os_version', 'weight',
       'price_euros'],
      dtype='object')


Now we convert `screen_size`, `ram`, `weight` and `price_euros` columns to numeric dtypes.
Whenever we convert a column from text to numeric data, we can follow this data cleaning workflow:

1. Explore data in column (use Series.unique() to view list of unique values)
2. Identify patterns and special cases
3. Remove non-digit characters
4. Convert to numeric dtype
5. Change column name if necessary 

* screen_size

In [4]:
print(laptops['screen_size'].dtype)
print(laptops['screen_size'].unique())

object
['13.3"' '15.6"' '15.4"' '14.0"' '12.0"' '11.6"' '17.3"' '10.1"' '13.5"'
 '12.5"' '13.0"' '18.4"' '13.9"' '12.3"' '17.0"' '15.0"' '14.1"' '11.3"']


The `screen_size` column values all contain a '"' non-numeric character which can be removed and the column should be renamed to maintain the units the values are recorded in. They do not represent integer values so should be converted to floats. 
Note: we will rename all columns simultaneously in the interests of efficiency after completing the data cleaning process   

In [5]:
#vectorize the removal of '"' character by using .str accessor and Series.str.replace() method and change the column dtype by using the Series.astype(dtype) method
laptops['screen_size'] = laptops['screen_size'].str.replace('"','').astype(float)
print(laptops['screen_size'].dtype)
print(laptops['screen_size'].unique())
#TODO: rename the column from 'screen_size' to 'screen_size_inches' 


float64
[13.3 15.6 15.4 14.  12.  11.6 17.3 10.1 13.5 12.5 13.  18.4 13.9 12.3
 17.  15.  14.1 11.3]


* ram

In [6]:
print(laptops['ram'].dtype)
print(laptops['ram'].unique())

object
['8GB' '16GB' '4GB' '2GB' '12GB' '6GB' '32GB' '24GB' '64GB']


In [7]:
#remove 'GB' substring and convert to int dtype
laptops['ram'] = laptops['ram'].str.replace('GB','').astype(int)
print(laptops['ram'].dtype)
print(laptops['ram'].unique())
#TODO: rename the column from 'ram' to 'ram_gb' 

int64
[ 8 16  4  2 12  6 32 24 64]


* weight

In [8]:
print(laptops['weight'].dtype)
print(laptops['weight'].unique())

object
['1.37kg' '1.34kg' '1.86kg' '1.83kg' '2.1kg' '2.04kg' '1.3kg' '1.6kg'
 '2.2kg' '0.92kg' '1.22kg' '0.98kg' '2.5kg' '1.62kg' '1.91kg' '2.3kg'
 '1.35kg' '1.88kg' '1.89kg' '1.65kg' '2.71kg' '1.2kg' '1.44kg' '2.8kg'
 '2kg' '2.65kg' '2.77kg' '3.2kg' '0.69kg' '1.49kg' '2.4kg' '2.13kg'
 '2.43kg' '1.7kg' '1.4kg' '1.8kg' '1.9kg' '3kg' '1.252kg' '2.7kg' '2.02kg'
 '1.63kg' '1.96kg' '1.21kg' '2.45kg' '1.25kg' '1.5kg' '2.62kg' '1.38kg'
 '1.58kg' '1.85kg' '1.23kg' '1.26kg' '2.16kg' '2.36kg' '2.05kg' '1.32kg'
 '1.75kg' '0.97kg' '2.9kg' '2.56kg' '1.48kg' '1.74kg' '1.1kg' '1.56kg'
 '2.03kg' '1.05kg' '4.4kg' '1.90kg' '1.29kg' '2.0kg' '1.95kg' '2.06kg'
 '1.12kg' '1.42kg' '3.49kg' '3.35kg' '2.23kg' '4.42kg' '2.69kg' '2.37kg'
 '4.7kg' '3.6kg' '2.08kg' '4.3kg' '1.68kg' '1.41kg' '4.14kg' '2.18kg'
 '2.24kg' '2.67kg' '2.14kg' '1.36kg' '2.25kg' '2.15kg' '2.19kg' '2.54kg'
 '3.42kg' '1.28kg' '2.33kg' '1.45kg' '2.79kg' '1.84kg' '2.6kg' '2.26kg'
 '3.25kg' '1.59kg' '1.13kg' '1.78kg' '1.10kg' '1.15kg' '1.27kg' 

In [9]:
#remove substrings 'kgs' and 'kg' and convert to float dtype
laptops['weight'] = laptops['weight'].str.replace('kgs','').str.replace('kg', '').astype(float)
print(laptops['weight'].dtype)
print(laptops['weight'].unique())
#TODO: rename the column from 'weight' to 'weight_kg' 

float64
[1.37  1.34  1.86  1.83  2.1   2.04  1.3   1.6   2.2   0.92  1.22  0.98
 2.5   1.62  1.91  2.3   1.35  1.88  1.89  1.65  2.71  1.2   1.44  2.8
 2.    2.65  2.77  3.2   0.69  1.49  2.4   2.13  2.43  1.7   1.4   1.8
 1.9   3.    1.252 2.7   2.02  1.63  1.96  1.21  2.45  1.25  1.5   2.62
 1.38  1.58  1.85  1.23  1.26  2.16  2.36  2.05  1.32  1.75  0.97  2.9
 2.56  1.48  1.74  1.1   1.56  2.03  1.05  4.4   1.29  1.95  2.06  1.12
 1.42  3.49  3.35  2.23  4.42  2.69  2.37  4.7   3.6   2.08  4.3   1.68
 1.41  4.14  2.18  2.24  2.67  2.14  1.36  2.25  2.15  2.19  2.54  3.42
 1.28  2.33  1.45  2.79  1.84  2.6   2.26  3.25  1.59  1.13  1.78  1.15
 1.27  1.43  2.31  1.16  1.64  2.17  1.47  3.78  1.79  0.91  1.99  4.33
 1.93  1.87  2.63  3.4   3.14  1.94  1.24  4.6   4.5   2.73  1.39  2.29
 2.59  2.94  1.14  3.8   3.31  1.09  3.21  1.19  1.98  1.17  4.36  1.71
 2.32  4.2   1.55  0.81  1.18  2.72  1.31  3.74  1.76  1.54  2.83  2.07
 2.38  3.58  1.08  2.75  2.99  1.11  2.09  4.    0.99  3.52

* price_euros

In [10]:
print(laptops['price_euros'].dtype)
print(laptops['price_euros'].unique()[:10])

object
['1339,69' '898,94' '575,00' '2537,45' '1803,60' '400,00' '2139,97'
 '1158,70' '1495,00' '770,00']


In [11]:
#replace ',' with '.' and convert to float dtype
laptops['price_euros'] = laptops['price_euros'].str.replace(',', '.').astype(float)
print(laptops['price_euros'].dtype)
print(laptops['price_euros'].unique()[:10])


float64
[1339.69  898.94  575.   2537.45 1803.6   400.   2139.97 1158.7  1495.
  770.  ]


Now we extract the screen resolution, the processor speed and the storage (hard drive) capacity from the `screen`, `cpu` and `storage` columns respectively and create new columns.

* Screen resolution

In [12]:
print(laptops['screen'].dtype)
print(laptops['screen'].unique())

object
['IPS Panel Retina Display 2560x1600' '1440x900' 'Full HD 1920x1080'
 'IPS Panel Retina Display 2880x1800' '1366x768'
 'IPS Panel Full HD 1920x1080' 'IPS Panel Retina Display 2304x1440'
 'IPS Panel Full HD / Touchscreen 1920x1080'
 'Full HD / Touchscreen 1920x1080' 'Touchscreen / Quad HD+ 3200x1800'
 'IPS Panel Touchscreen 1920x1200' 'Touchscreen 2256x1504'
 'Quad HD+ / Touchscreen 3200x1800' 'IPS Panel 1366x768'
 'IPS Panel 4K Ultra HD / Touchscreen 3840x2160'
 'IPS Panel Full HD 2160x1440' '4K Ultra HD / Touchscreen 3840x2160'
 'Touchscreen 2560x1440' '1600x900' 'IPS Panel 4K Ultra HD 3840x2160'
 '4K Ultra HD 3840x2160' 'Touchscreen 1366x768'
 'IPS Panel Full HD 1366x768' 'IPS Panel 2560x1440'
 'IPS Panel Full HD 2560x1440' 'IPS Panel Retina Display 2736x1824'
 'Touchscreen 2400x1600' '2560x1440' 'IPS Panel Quad HD+ 2560x1440'
 'IPS Panel Quad HD+ 3200x1800'
 'IPS Panel Quad HD+ / Touchscreen 3200x1800'
 'IPS Panel Touchscreen 1366x768' '1920x1080'
 'IPS Panel Full HD 1920x120

In [0]:
#split each entry using single whitespace character as delimiter and extract last element from list of constituents
#assign first last element of list of constituents to new column 'screen_res'
#join rest of elements of list of constituents into a single string and assign back to 'screen' column
screen_split = laptops['screen'].str.split()
laptops['screen_res'] = screen_split.str[-1]
laptops['screen'] = screen_split.str[:-1].str.join(' ')
#TODO: rename 'screen' column as 'screen_type'

In [14]:
print(laptops['screen'].dtype)
print(laptops['screen'].unique())
print()
print(laptops['screen_res'].dtype)
print(laptops['screen_res'].unique())

object
['IPS Panel Retina Display' '' 'Full HD' 'IPS Panel Full HD'
 'IPS Panel Full HD / Touchscreen' 'Full HD / Touchscreen'
 'Touchscreen / Quad HD+' 'IPS Panel Touchscreen' 'Touchscreen'
 'Quad HD+ / Touchscreen' 'IPS Panel'
 'IPS Panel 4K Ultra HD / Touchscreen' '4K Ultra HD / Touchscreen'
 'IPS Panel 4K Ultra HD' '4K Ultra HD' 'IPS Panel Quad HD+'
 'IPS Panel Quad HD+ / Touchscreen' 'IPS Panel Touchscreen / 4K Ultra HD'
 'Touchscreen / Full HD' 'Quad HD+' 'Touchscreen / 4K Ultra HD']

object
['2560x1600' '1440x900' '1920x1080' '2880x1800' '1366x768' '2304x1440'
 '3200x1800' '1920x1200' '2256x1504' '3840x2160' '2160x1440' '2560x1440'
 '1600x900' '2736x1824' '2400x1600']


Note: Some of the entries in 'screen' column are now empty strings - these will need to be dealt with when we manage null values later in this workflow. 

* Processor speed

In [15]:
print(laptops['cpu'].dtype)
print(laptops['cpu'].unique())

object
['Intel Core i5 2.3GHz' 'Intel Core i5 1.8GHz'
 'Intel Core i5 7200U 2.5GHz' 'Intel Core i7 2.7GHz'
 'Intel Core i5 3.1GHz' 'AMD A9-Series 9420 3GHz' 'Intel Core i7 2.2GHz'
 'Intel Core i7 8550U 1.8GHz' 'Intel Core i5 8250U 1.6GHz'
 'Intel Core i3 6006U 2GHz' 'Intel Core i7 2.8GHz'
 'Intel Core M m3 1.2GHz' 'Intel Core i7 7500U 2.7GHz'
 'Intel Core i7 2.9GHz' 'Intel Core i3 7100U 2.4GHz'
 'Intel Atom x5-Z8350 1.44GHz' 'Intel Core i5 7300HQ 2.5GHz'
 'AMD E-Series E2-9000e 1.5GHz' 'Intel Core i5 1.6GHz'
 'Intel Core i7 8650U 1.9GHz' 'Intel Atom x5-Z8300 1.44GHz'
 'AMD E-Series E2-6110 1.5GHz' 'AMD A6-Series 9220 2.5GHz'
 'Intel Celeron Dual Core N3350 1.1GHz' 'Intel Core i3 7130U 2.7GHz'
 'Intel Core i7 7700HQ 2.8GHz' 'Intel Core i5 2.0GHz'
 'AMD Ryzen 1700 3GHz' 'Intel Pentium Quad Core N4200 1.1GHz'
 'Intel Atom x5-Z8550 1.44GHz' 'Intel Celeron Dual Core N3060 1.6GHz'
 'Intel Core i5 1.3GHz' 'AMD FX 9830P 3GHz' 'Intel Core i7 7560U 2.4GHz'
 'AMD E-Series 6110 1.5GHz' 'Intel Core

On inspection: all processor speeds are quoted at the end of the entry in the `cpu` column. All processor speeds are quoted in GHz. Processor speed is not an integer value so will need to be converted to a float dtype.

In [0]:
cpu_split = laptops['cpu'].str.split()
laptops['cpu_speed_ghz'] = cpu_split.str[-1].str.replace('GHz', '').astype(float) #remove 'GHz' substring and convert to float dtype
laptops['cpu'] = cpu_split.str[:-1].str.join(' ')
#TODO: rename 'cpu' column as 'cpu_type'

In [17]:
print(laptops['cpu'].dtype)
print(laptops['cpu'].unique()[:10])
print()
print(laptops['cpu_speed_ghz'].dtype)
print(laptops['cpu_speed_ghz'].unique())

object
['Intel Core i5' 'Intel Core i5 7200U' 'Intel Core i7'
 'AMD A9-Series 9420' 'Intel Core i7 8550U' 'Intel Core i5 8250U'
 'Intel Core i3 6006U' 'Intel Core M m3' 'Intel Core i7 7500U'
 'Intel Core i3 7100U']

float64
[2.3  1.8  2.5  2.7  3.1  3.   2.2  1.6  2.   2.8  1.2  2.9  2.4  1.44
 1.5  1.9  1.1  1.3  2.6  3.6  3.2  1.   2.1  0.9  1.92]


* Storage capacity

In [18]:
print(laptops['storage'].dtype)
print(laptops['storage'].unique())

object
['128GB SSD' '128GB Flash Storage' '256GB SSD' '512GB SSD' '500GB HDD'
 '256GB Flash Storage' '1TB HDD' '32GB Flash Storage'
 '128GB SSD +  1TB HDD' '256GB SSD +  256GB SSD' '64GB Flash Storage'
 '256GB SSD +  1TB HDD' '256GB SSD +  2TB HDD' '32GB SSD' '2TB HDD'
 '64GB SSD' '1TB Hybrid' '512GB SSD +  1TB HDD' '1TB SSD'
 '256GB SSD +  500GB HDD' '128GB SSD +  2TB HDD' '512GB SSD +  512GB SSD'
 '16GB SSD' '16GB Flash Storage' '512GB SSD +  256GB SSD'
 '512GB SSD +  2TB HDD' '64GB Flash Storage +  1TB HDD' '1GB SSD'
 '1TB HDD +  1TB HDD' '32GB HDD' '1TB SSD +  1TB HDD'
 '512GB Flash Storage' '128GB HDD' '240GB SSD' '8GB SSD' '508GB Hybrid'
 '512GB SSD +  1TB Hybrid' '256GB SSD +  1TB Hybrid']


The entries in `storage` column show there are four types of storage: 'SSD', 'HDD', 'Hybrid' and 'Flash Storage'. Note: if an entry contains more than one type of storage it contains a '+' character. We will use the existing `storage` column to create four new columns: `storage_1_type`, `storage_2_type` and `storage_1_size_gb`, `storage_2_size_gb` and then remove the `storage` column. Finally, we will add a new column `total_storage_gb` which will be the sum for each row of `storage_1_size_gb` and `storage_2_size_gb` entries in GB. 



In [0]:
#create bool series from storage_split series: true for entries with 2 storage types
two_types_bool = laptops['storage'].str.contains('+', regex=False)

#assign entries with one type of storage to new columns
laptops.loc[~two_types_bool, 'storage_1_type'] = (laptops.loc[~two_types_bool, 'storage'].
                          str.split(n=1).
                          str[-1].
                          str.strip().
                          str.lower()
)

laptops.loc[~two_types_bool, 'storage_1_size_gb'] = (laptops.loc[~two_types_bool, 'storage'].
                             str.split(n=1).
                             str[0].
                             str.replace('GB', '').
                             str.strip()
)

#assign entries with two types of storage to new and existing columns
laptops.loc[two_types_bool, 'storage_1_type'] = (laptops.loc[two_types_bool, 'storage'].
                          str.split('+').
                          str[0].
                          str.split(n=1).
                          str[-1].
                          str.strip().
                          str.lower()
)

laptops.loc[two_types_bool, 'storage_2_type'] = (laptops.loc[two_types_bool, 'storage'].
                          str.split('+').
                          str[-1].
                          str.split(n=1).
                          str[-1].
                          str.strip().
                          str.lower()
)

laptops.loc[two_types_bool, 'storage_1_size_gb'] = (laptops.loc[two_types_bool, 'storage'].
                          str.split('+').
                          str[0].
                          str.split(n=1).
                          str[0].
                          str.replace('GB', '').
                          str.strip()
)

laptops.loc[two_types_bool, 'storage_2_size_gb'] = (laptops.loc[two_types_bool, 'storage'].
                          str.split('+').
                          str[-1].
                          str.split(n=1).
                          str[0].
                          str.replace('GB', '').
                          str.strip()
)

                                      

In [0]:
#convert TB storage sizes to GB 
laptops.loc[laptops['storage_1_size_gb'].str.contains('TB'), 'storage_1_size_gb'] = (laptops.loc[laptops['storage_1_size_gb'].str.contains('TB'), 'storage_1_size_gb'].
                                                                                 str.replace('TB', '').
                                                                                 astype(int)
) * 1000

laptops.loc[laptops['storage_2_size_gb'].str.contains('TB', na=False), 'storage_2_size_gb'] = (laptops.loc[laptops['storage_2_size_gb'].str.contains('TB', na=False), 'storage_2_size_gb'].
                                                                                 str.replace('TB', '').
                                                                                 astype(int)
) * 1000

#remove 'storage' column from laptops dataframe
laptops.drop(columns=['storage'], inplace=True)

In [0]:
#cast storage size columns to float dtypes 
laptops['storage_1_size_gb'] = laptops['storage_1_size_gb'].astype(float)
laptops['storage_2_size_gb'] = laptops['storage_2_size_gb'].astype(float)
#add 'total_storage_gb' column taking NaN values in 'storage_2_size_gb' column as 0. cast 'total_storage_gb' column as int type
laptops['total_storage_gb'] = laptops['storage_1_size_gb'] + laptops['storage_2_size_gb'].fillna(0)
laptops['total_storage_gb'] = laptops['total_storage_gb'].astype(int)








In [22]:
#explore new columns in laptops df
print(laptops.columns)
print()
print(laptops['storage_1_type'].value_counts(dropna=False))
print(laptops['storage_1_type'].dtype)
print()
print(laptops['storage_2_type'].value_counts(dropna=False))
print(laptops['storage_2_type'].dtype)
print()
print(laptops['storage_1_size_gb'].value_counts(dropna=False))
print(laptops['storage_1_size_gb'].dtype)
print()
print(laptops['storage_2_size_gb'].value_counts(dropna=False))
print(laptops['storage_2_size_gb'].dtype)
print()
print(laptops['total_storage_gb'].value_counts(dropna=False))
print(laptops['total_storage_gb'].dtype)

Index(['manufacturer', 'model_name', 'category', 'screen_size', 'screen',
       'cpu', 'ram', 'gpu', 'os', 'os_version', 'weight', 'price_euros',
       'screen_res', 'cpu_speed_ghz', 'storage_1_type', 'storage_1_size_gb',
       'storage_2_type', 'storage_2_size_gb', 'total_storage_gb'],
      dtype='object')

ssd              843
hdd              375
flash storage     75
hybrid            10
Name: storage_1_type, dtype: int64
object

NaN       1095
hdd        202
ssd          4
hybrid       2
Name: storage_2_type, dtype: int64
object

256.0     508
1000.0    250
128.0     177
512.0     140
500.0     132
32.0       45
64.0       17
2000.0     16
16.0       10
1.0         5
508.0       1
8.0         1
240.0       1
Name: storage_1_size_gb, dtype: int64
float64

NaN       1095
1000.0     187
2000.0      15
256.0        3
500.0        2
512.0        1
Name: storage_2_size_gb, dtype: int64
float64

256     420
1000    247
500     132
512     122
1128     94
128      81
1256     74
32    

Now rename all columns clearly and succinctly, including units where appropriate. 



In [23]:
#TODO: rename 'cpu' column as 'cpu_type'
#TODO: rename 'screen' column as 'screen_type'
#TODO: rename the column from 'weight' to 'weight_kg' 
#TODO: rename the column from 'ram' to 'ram_gb' 
#TODO: rename the column from 'screen_size' to 'screen_size_inches'

laptops.rename(columns={'cpu':'cpu_type', 'screen':'screen_type', 'weight':'weight_kg', 'ram':'ram_gb', 'screen_size':'screen_size_inches'}, inplace=True)
print(laptops.columns)


Index(['manufacturer', 'model_name', 'category', 'screen_size_inches',
       'screen_type', 'cpu_type', 'ram_gb', 'gpu', 'os', 'os_version',
       'weight_kg', 'price_euros', 'screen_res', 'cpu_speed_ghz',
       'storage_1_type', 'storage_1_size_gb', 'storage_2_type',
       'storage_2_size_gb', 'total_storage_gb'],
      dtype='object')


In [24]:
#check null entries in each column of laptops df
print(laptops.isnull().sum())

manufacturer             0
model_name               0
category                 0
screen_size_inches       0
screen_type              0
cpu_type                 0
ram_gb                   0
gpu                      0
os                       0
os_version             170
weight_kg                0
price_euros              0
screen_res               0
cpu_speed_ghz            0
storage_1_type           0
storage_1_size_gb        0
storage_2_type        1095
storage_2_size_gb     1095
total_storage_gb         0
dtype: int64


Looks good! Getting there!

There are a few options for handling missing values:

1. Remove any rows that have missing values.
2. Remove any columns that have missing values.
3. Fill the missing values with some other value.
4. Leave the missing values as is.

We do not want to remove any rows containing null values in the `storage_2_type` or the `storage_2_size_gb` columns because we know these rows contain data in the other storage related columns. We have decided to leave the missing values as is for these columns rather than replace with some other value. Note: we will need to remove these missing values if we wish to pass the data to a ML algorithm. 

It's now clear that we have only only one other column with null values, `os_version`, which has 170 missing values. Lets explore the `os` and `os_version` columns in the laptops df:



In [25]:
print(laptops['os_version'].value_counts(dropna=False))
print()
print(laptops['os'].value_counts(dropna=False))


10      1072
NaN      170
7         45
10 S       8
X          8
Name: os_version, dtype: int64

Windows      1125
No OS          66
Linux          62
Chrome OS      27
macOS          13
Mac OS          8
Android         2
Name: os, dtype: int64


There are 66 rows with 'No OS' value for the `os` column. We will now look at the `os` column values only for rows which have a missing entry for 'os_version`:  

In [26]:
null_os_version_bool = laptops['os_version'].isnull()
print(laptops.loc[null_os_version_bool, 'os'].value_counts())

No OS        66
Linux        62
Chrome OS    27
macOS        13
Android       2
Name: os, dtype: int64


Two observations:

1. For rows which have an entry of 'No OS' in the `os` column there should be no entry in the `os_version` column. We will insert 'Version Unknown' for these rows. 

2. There are 13 rows with a missing entry in the `os_version` column with a value of 'macOS' in the `os` column. Because we know and love Apple, we can enter 'X' in the `os_version` entry for these rows!

3. There are 'Mac OS' and 'macOS' entries in the `os` column. We will convert 'macOS' values to 'Mac OS'. 


In [27]:
#observe entries in 'os_version' column for rows with 'No OS' entry in 'os' column
print(laptops.loc[laptops.loc[:, 'os'] == 'No OS', 'os_version'].value_counts(dropna=False))
#replace missing entries in 'os_version' column with 'Version Unknown' 
laptops.loc[(laptops['os'] == 'No OS') | (laptops['os_version'].isnull()), 'os_version'] = 'Version Unknown'
print()
print(laptops.loc[laptops.loc[:, 'os'] == 'No OS', 'os_version'].value_counts(dropna=False))


NaN    66
Name: os_version, dtype: int64

Version Unknown    66
Name: os_version, dtype: int64


In [0]:
laptops.loc[laptops["os"] == "macOS", "os_version"] = "X" 
laptops.loc[laptops["os"] == "macOS", 'os'] = 'Mac OS'

In [29]:
print(laptops['os_version'].value_counts(dropna=False))
print()
print(laptops['os'].value_counts(dropna=False))



10                 1072
Version Unknown     157
7                    45
X                    21
10 S                  8
Name: os_version, dtype: int64

Windows      1125
No OS          66
Linux          62
Chrome OS      27
Mac OS         21
Android         2
Name: os, dtype: int64


No missing values remain in either column. Now to quickly check the other changes we have made:

In [30]:
#observe entries in 'os_version' column for rows with 'Mac OS' in 'os' column
laptops.loc[laptops['os'] == 'Mac OS', 'os_version'].value_counts(dropna=False)

X    21
Name: os_version, dtype: int64

We now have a much cleaner dataframe and time for a bit of analysis. 

## Analysis of cleaned data

We will address the following questions in order:

1. Are laptops made by Apple more expensive than those made by other manufacturers?
2.  What is the best value laptop with a screen size of 15" or more?
3.  Which laptop has the most storage space?

## 1. Are laptops made by Apple more expensive than those made by other manufacturers?

In [31]:
#find mean value of 'price_euros' column entries for laptops from each manufacturer
average_price_by_manufacturer = {}

for m in laptops['manufacturer'].unique():
  selected_rows_price_euros = laptops.loc[laptops['manufacturer'] == m, 'price_euros']
  average_price_by_manufacturer[m] = selected_rows_price_euros.mean()

#sort the resultant dictionary by value
print(sorted(average_price_by_manufacturer.items(), key=lambda x: x[1], reverse=True))
print(len(average_price_by_manufacturer))



[('Razer', 3346.1428571428573), ('LG', 2099.0), ('MSI', 1728.9081481481485), ('Google', 1677.6666666666667), ('Microsoft', 1612.3083333333334), ('Apple', 1564.1985714285713), ('Huawei', 1424.0), ('Samsung', 1413.4444444444443), ('Toshiba', 1267.8125), ('Dell', 1186.0689898989892), ('Xiaomi', 1133.4625), ('Asus', 1104.1693670886075), ('Lenovo', 1086.3844444444444), ('HP', 1067.774854014598), ('Fujitsu', 729.0), ('Acer', 626.7758252427185), ('Chuwi', 314.2966666666667), ('Mediacom', 295.0), ('Vero', 217.425)]
19


It appears that on average the laptops in the dataset made by 'Razer', 'LG', 'MSI', 'Google' and 'Microsoft' are more expensive than those made by 'Apple'. Apple is in 6th position out of 19 so it is one of the more expensive manufacturers of laptops. 

## 2.  What is the best value laptop with a screen size of 15" or more?

In [32]:
#filter rows by screen size >= 15 inches
large_screens = laptops.loc[laptops['screen_size_inches'] >= 15, ['manufacturer', 'model_name', 'screen_size_inches', 'price_euros']]
print(large_screens.sort_values(by=['price_euros']).iloc[0])

manufacturer                          Acer
model_name            Chromebook C910-C2ST
screen_size_inches                    15.6
price_euros                            199
Name: 290, dtype: object


The best value laptop with a screen size of 15" or more is 'Chromebook C910-C2ST' by 'Acer'. 

## 3. Which laptop has the most storage space?


In [42]:
print(laptops.loc[:, ['manufacturer', 'model_name', 'total_storage_gb']].sort_values(by=['total_storage_gb'], ascending=False).iloc[0])
print()
print(laptops.loc[laptops['total_storage_gb'] == 2512, ['manufacturer', 'model_name', 'total_storage_gb']])

manufacturer                   MSI
model_name          GS73VR Stealth
total_storage_gb              2512
Name: 894, dtype: object

    manufacturer                         model_name  total_storage_gb
370         Asus                       ZenBook Flip              2512
894          MSI                     GS73VR Stealth              2512
977         Asus  Q534UX-BHI7T19 (i7-7500U/16GB/2TB              2512


There are three laptops with 2512GB of storage.

# Export the cleaned dataframe to a csv file



In [0]:
laptops.to_csv('/content/drive/My Drive/Datasets/laptops_cleaned.csv', index=False)